torchx/schedulers/ray_scheduler.py

```diff
@@ -4,236 +4,323 @@
+        return _has_ray
+
+    @dataclass
+    class RayJob:
+        """Represents a job that should be run on a Ray cluster.
+
+        Attributes:
+            app_id:
+                The unique ID of the application (a.k.a. job).
+            cluster_config_file:
+                The Ray cluster configuration file.
+            cluster_name:
+                The cluster name to use.
+            dashboard_address:
+                The existing dashboard IP address to connect to
+            working_dir:
+            The working directory to copy to the cluster
+            requirements:
+                The libraries to install on the cluster per requirements.txt
+            actors:
+                The Ray actors which represent the job to be run. This attribute is
+                dumped to a JSON file and copied to the cluster where `ray_main.py`
+                uses it to initiate the job.
+        """
+
+        app_id: str
+        cluster_config_file: Optional[str] = None
+        cluster_name: Optional[str] = None
+        dashboard_address: Optional[str] = None
+        working_dir: Optional[str] = None
+        requirements: Optional[str] = None
+        scripts: Set[str] = field(default_factory=set)
+        actors: List[RayActor] = field(default_factory=list)
+
+    class RayScheduler(Scheduler):
+        def __init__(self, session_name: str) -> None:
+            super().__init__("ray", session_name)
+
+        # TODO: Add address as a potential CLI argument after writing ray.status() or passing in config file
+        def run_opts(self) -> runopts:
+            opts = runopts()
+            opts.add(
+                "cluster_config_file",
+                type_=str,
+                required=False,
+                help="Use CLUSTER_CONFIG_FILE to access or create the Ray cluster.",
+            )
+            opts.add(
+                "cluster_name",
+                type_=str,
+                help="Override the configured cluster name.",
+            )
+            opts.add(
+                "dashboard_address",
+                type_=str,
+                required=False,
+                default="127.0.0.1:8265",
+                help="Use ray status to get the dashboard address you will submit jobs against",
+            )
+            opts.add(
+                "working_dir",
+                type_=str,
+                help="Copy the the working directory containing the Python scripts to the cluster.",
+            )
+            opts.add("requirements", type_=str, help="Path to requirements.txt")
+            return opts
+
+        def schedule(self, dryrun_info: AppDryRunInfo[RayJob]) -> str:
+            cfg: RayJob = dryrun_info.request
+
+            # Create serialized actors for ray_driver.py
+            actors = cfg.actors
+            dirpath = mkdtemp()
+            serialize(actors, dirpath)
+
+            ip_address = "127.0.0.1:8265"
```

Comment:
remove this?

Reply:
this is primarily to make pyre happy since only one of the two args are necessary: dashboard_url or cluster_yaml but pyre doesn't know which